src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -0,0 +1,430 @@
+  const std::lock_guard l(lfuda_lock);
+  if (entries_heap.empty())
+    return nullptr;
+
+  /* Get victim cache block */
+  std::string key = entries_heap.top()->key;
+  CacheBlock* victim = new CacheBlock();
+
+  victim->cacheObj.bucketName = key.substr(0, key.find('_')); 
+  key.erase(0, key.find('_') + 1);
+  victim->cacheObj.objName = key.substr(0, key.find('_'));
+  victim->blockID = entries_heap.top()->offset;
+  victim->size = entries_heap.top()->len;
+
+  if (dir->get(victim, y) < 0) {
+    return nullptr;
+  }
+
+  return victim;
+}
+
+void LFUDAPolicy::shutdown() {
+  dir->shutdown();
+  
+  // call cancel() on the connection's executor
+  boost::asio::dispatch(conn->get_executor(), [c = conn] { c->cancel(); });
+}
+
+int LFUDAPolicy::exist_key(std::string key) {
+  if (entries_map.count(key) != 0) {
+    return true;
+  }
+
+  return false;
+}
+
+#if 0
+int LFUDAPolicy::get_block(const DoutPrefixProvider* dpp, CacheBlock* block, rgw::cache::CacheDriver* cacheDriver, optional_yield y) {
+  response<std::string> resp;
+  int age = get_age(y);
+
+  if (exist_key(build_index(block->cacheObj.bucketName, block->cacheObj.objName, block->blockID, block->size))) { /* Local copy */
+    auto it = entries_map.find(build_index(block->cacheObj.bucketName, block->cacheObj.objName, block->blockID, block->size));
+    it->second->localWeight += age;
+    return cacheDriver->set_attr(dpp, block->cacheObj.objName, "localWeight", std::to_string(it->second->localWeight), y);
+  } else {
+    if (eviction(dpp, block->size, cacheDriver, y) < 0)
+      return -1; 
+
+    int exists = dir->exist_key(block, y);
+    if (exists > 0) { /* Remote copy */
+      if (dir->get(block, y) < 0) {
+	return -1;
+      } else {
+	if (!block->hostsList.empty()) { 
+	  block->globalWeight += age;
+	  
+	  if (dir->update_field(block, "globalWeight", std::to_string(block->globalWeight), y) < 0) {
+	    return -1;
+	  } else {
+	    return 0;
+	  }
+	} else {
+          return -1;
+        }
+      }
+    } else if (!exists) { /* No remote copy */
+      // localWeight += age;
+      //return cacheDriver->set_attr(dpp, block->cacheObj.objName, "localWeight", std::to_string(it->second->localWeight), y);
+      return 0;
+    } else {
+      return -1;
+    }
+  }
+}
+#endif
+
+int LFUDAPolicy::eviction(const DoutPrefixProvider* dpp, uint64_t size, optional_yield y) {
+  uint64_t freeSpace = cacheDriver->get_free_space(dpp);
+
+  while (freeSpace < size) { // TODO: Think about parallel reads and writes; can this turn into an infinite loop? 
+    CacheBlock* victim = find_victim(dpp, y);
+
+    if (victim == nullptr) {
+      ldpp_dout(dpp, 10) << "LFUDAPolicy::" << __func__ << "(): Could not retrieve victim block." << dendl;
+      delete victim;
+      return -1;
+    }
+
+    std::string key = victim->cacheObj.bucketName + "_" + victim->cacheObj.objName + "_" + std::to_string(victim->blockID) + "_" + std::to_string(victim->size);
```

Comment:
@samarahu : I noticed that the `key` here doesn't match with the way `key` that is passed in the `update` method, so this will always result in the key not being found. P.S: The key names now include `version` also.